src/worker/tasks/nonceHealthCheckWorker.ts

```diff
@@ -0,0 +1,165 @@
+
+// Configuration
+
+// Number of consecutive periods to check
+// Checking over multiple periods helps to avoid false positives due to timing issues
+const CHECK_PERIODS = 3;
+
+// Frequency of the worker
+const RUN_FREQUENCY_SECONDS = 60; // Run every minute
+
+// Interfaces
+interface NonceState {
+  onchainNonce: number;
+  largestSentNonce: number;
+}
+
+interface WalletHealth {
+  walletAddress: Address;
+  chainId: number;
+  isStuck: boolean;
+  onchainNonce: number;
+  largestSentNonce: number;
+}
+
+// Initialize the worker
+export const initNonceHealthCheckWorker = () => {
+  NonceHealthCheckQueue.q.add("cron", "", {
+    repeat: { pattern: `*/${RUN_FREQUENCY_SECONDS} * * * * *` },
+    jobId: "nonce-health-check-cron",
+  });
+
+  const _worker = new Worker(NonceHealthCheckQueue.q.name, handler, {
+    connection: redis,
+    concurrency: 1,
+  });
+  logWorkerExceptions(_worker);
+};
+
+// Main handler function
+const handler: Processor<any, void, string> = async (job: Job<string>) => {
+  const allWallets = await getUsedBackendWallets();
+
+  const walletHealthPromises = allWallets.map(
+    async ({ chainId, walletAddress }) => {
+      const [_, isStuck, currentState] = await Promise.all([
+        updateNonceHistory(walletAddress, chainId),
+        isQueueStuck(walletAddress, chainId),
+        getCurrentNonceState(walletAddress, chainId),
+      ]);
+
+      return {
+        walletAddress,
+        chainId,
+        isStuck,
+        onchainNonce: currentState.onchainNonce,
+        largestSentNonce: currentState.largestSentNonce,
+      };
+    },
+  );
+
+  const walletHealthResults = await Promise.all(walletHealthPromises);
+  logWalletHealth(walletHealthResults, job);
+};
+
+// Check if a queue is stuck
+async function isQueueStuck(
+  walletAddress: Address,
+  chainId: number,
+): Promise<boolean> {
+  const historicalStates = await getHistoricalNonceStates(
+    walletAddress,
+    chainId,
+    CHECK_PERIODS,
+  );
+
+  // ensure we have enough data to check
+  if (historicalStates.length < CHECK_PERIODS) return false;
+
+  // if for every period, the onchain nonce has not changed, and the internal nonce has strictly increased
+  // then the queue is stuck
+  const isStuckForAllPeriods = historicalStates.every((state, index) => {
+    if (index === historicalStates.length - 1) return true; // Last (oldest) state
+
+    const prevState = historicalStates[index + 1];
+    return (
+      state.onchainNonce === prevState.onchainNonce &&
+      state.largestSentNonce > prevState.largestSentNonce
+    );
+  });
+
+  return isStuckForAllPeriods;
+}
+
+// Get current nonce state
+async function getCurrentNonceState(
+  walletAddress: Address,
+  chainId: number,
+): Promise<NonceState> {
+  const [onchainNonce, largestSentNonce] = await Promise.all([
+    getLastUsedOnchainNonce(chainId, walletAddress),
+    inspectNonce(chainId, walletAddress),
+  ]);
+
+  return {
+    onchainNonce: parseInt(onchainNonce.toString()),
```

Comment:
Isn't `onchainNonce` is already a number? This line stringifies it, then parses int again?

Reply:
bump on this comment